src/musig.rs

```diff
@@ -0,0 +1,1286 @@
+    /// # let pub_key2 = PublicKey::from_secret_key(&secp, &sk2);
+    /// #
+    /// let mut key_agg_cache = MusigKeyAggCache::new(&secp, &[pub_key1, pub_key2]);
+    ///
+    /// let tweak = SecretKey::from_slice(b"Insecure tweak, Don't use this!!").unwrap(); // tweak could be from BIP32
+    /// let _tweaked_key = key_agg_cache.pubkey_ec_tweak_add(&secp, tweak).unwrap();
+    /// # }
+    /// ```
+    pub fn pubkey_ec_tweak_add<C: Verification>(
+        &mut self,
+        secp: &Secp256k1<C>,
+        tweak: SecretKey,
+    ) -> Result<PublicKey, MusigTweakErr> {
+        let cx = secp.ctx().as_ptr();
+        unsafe {
+            let mut out = PublicKey::from(ffi::PublicKey::new());
+            if ffi::secp256k1_musig_pubkey_ec_tweak_add(
+                cx,
+                out.as_mut_c_ptr(),
+                self.as_mut_ptr(),
+                tweak.as_c_ptr(),
+            ) == 0
+            {
+                Err(MusigTweakErr::InvalidTweak)
+            } else {
+                self.1 = out.x_only_public_key().0;
+                Ok(out)
+            }
+        }
+    }
+
+    /// Apply "x-only" tweaking to a public key in a [`MusigKeyAggCache`].
+    ///
+    /// This is done by adding the generator multiplied with `tweak32` to it. Returns the tweaked [`XOnlyPublicKey`].
+    /// This is useful in creating taproot outputs.
+    /// This function is required if you want to _sign_ for a tweaked aggregate key.
+    ///
+    /// # Arguments:
+    ///
+    /// * `secp` : [`Secp256k1`] context object initialized for verification
+    /// * `tweak`: tweak of type [`SecretKey`] with which to tweak the aggregated key
+    ///
+    /// # Errors:
+    ///
+    /// If resulting public key would be invalid (only when the tweak is the negation of the corresponding
+    /// secret key). For uniformly random 32-byte arrays(for example, in BIP341 taproot tweaks) the chance of
+    /// being invalid is negligible (around 1 in 2^128)
+    ///
+    /// Example:
+    ///
+    /// ```rust
+    /// # # [cfg(any(test, feature = "rand-std"))] {
+    /// # use secp256k1_zkp::rand::{thread_rng, RngCore};
+    /// # use secp256k1_zkp::{MusigKeyAggCache, Secp256k1, SecretKey, Keypair, PublicKey};
+    /// # let secp = Secp256k1::new();
+    /// # let sk1 = SecretKey::new(&mut thread_rng());
+    /// # let pub_key1 = PublicKey::from_secret_key(&secp, &sk1);
+    /// # let sk2 = SecretKey::new(&mut thread_rng());
+    /// # let pub_key2 = PublicKey::from_secret_key(&secp, &sk2);
+    ///
+    /// let mut key_agg_cache = MusigKeyAggCache::new(&secp, &[pub_key1, pub_key2]);
+    ///
+    /// let tweak = SecretKey::from_slice(b"Insecure tweak, Don't use this!!").unwrap(); // tweak could be from tap
+    /// let _x_only_key_tweaked = key_agg_cache.pubkey_xonly_tweak_add(&secp, tweak).unwrap();
+    /// # }
+    /// ```
+    pub fn pubkey_xonly_tweak_add<C: Verification>(
+        &mut self,
+        secp: &Secp256k1<C>,
+        tweak: SecretKey,
```

Comment:
`Scalar`

Reply:
Done in https://github.com/rust-bitcoin/rust-secp256k1/pull/716/commits/379f70a152c5be546c678b4179b2ef99c7d955fb